### PR TITLE
Add validator to check consistency of resource name used for IamResource

### PR DESCRIPTION
**Pull request by hpmellema**
### Background
The [`IamResource`](https://smithy.io/2.0/aws/aws-iam.html#aws-iam-iamresource-trait) trait indicates properties of a Smithy resource in AWS IAM. If no `name` parameter is provided to the `IamResource` trait then a shape's IamResource name is assumed to just be the resource's name, otherwise the `name` parameter is used.

The name of an IamResource is expected to match the name of the resource in the resource's ARN. For examples if the resource has an arn trait of `@arn(template: "my-shape")` then the `IamResource` name should be `"my-shape"` for that resource.

### Description of changes:
- Adds new validator and associated unit tests to check that the name specified for the `IamResource` trait is consistent with the resource name in the resource's ARN template

Note: This validator throws a DANGER event instead of an ERROR so that the validation event can be suppressed if needed for backwards compatibility.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-aws-iam-traits/src/main/resources/META-INF/services/software.amazon.smithy.model.validation.Validator [hunk 1]
```diff
@@ -1 +1,2 @@
 software.amazon.smithy.aws.iam.traits.ConditionKeysValidator
+software.amazon.smithy.aws.iam.traits.IamResourceTraitValidator
```
Comment (mtdowling) [inline smithy-aws-iam-traits/src/main/resources/META-INF/services/software.amazon.smithy.model.validation.Validator:2]: nit - missing newline

File: smithy-aws-iam-traits/src/test/java/software/amazon/smithy/aws/iam/traits/IamResourceValidatorTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+public class IamResourceValidatorTest {
```
Comment (mtdowling) [inline smithy-aws-iam-traits/src/test/java/software/amazon/smithy/aws/iam/traits/IamResourceValidatorTest.java:19]: Can we take this as an opportunity to add a test runner for .smithy/.errors?

File: smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                            + "`%s`, for that same resource.",
+                            resourceName, arnTrait.getTemplate())));
+                }
+            }
+        }
+        return results;
+    }
+
+    private List<String> parseArnComponents(String arnTemplate) {
+        return new ArrayList<>(Arrays.asList(arnTemplate.split("/")));
```
Comment (mtdowling) [inline smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java:60]: Returning Arrays.asList's result directly doesn't work?

File: smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+public class IamResourceTraitValidator extends AbstractValidator {
+    @Override
+    public List<ValidationEvent> validate(Model model) {
+        List<ValidationEvent> results = new ArrayList<>();
+        for (ResourceShape resource : model.getResourceShapesWithTrait(IamResourceTrait.class)) {
+            // If the resource has both the IamResourceTrait and Arn trait,
+            // check that the resource name is consistent between the two traits
+            if (resource.hasTrait(ArnTrait.class)) {
+                String resourceName = resource.expectTrait(IamResourceTrait.class).getName()
+                    .orElse(StringUtils.lowerCase(resource.getId().getName()));
```
Comment (mtdowling) [inline smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java:44]: Use orElseGet when doing a computation like this

File: smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+                ArnTrait arnTrait = resource.expectTrait(ArnTrait.class);
+                List<String> arnComponents = parseArnComponents(arnTrait.getTemplate());
+                if (!arnComponents.contains(resourceName)) {
+                    results.add(danger(resource, String.format(
+                            "The `@aws.iam#iamResource trait applied to the resource "
```
Comment (mtdowling) [inline smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java:49]:
```suggestion
                            "The `@aws.iam#iamResource` trait applied to this resource "
```

File: smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                            + "defines a resource name, `%s`, that does not match the `@arn` template, "
```
Comment (mtdowling) [inline smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java:50]: Can we be more specific here and say an IAM resource name?

File: smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                            + "`%s`, for that same resource.",
```
Comment (mtdowling) [inline smithy-aws-iam-traits/src/main/java/software/amazon/smithy/aws/iam/traits/IamResourceTraitValidator.java:51]:
```suggestion
                            + "`%s`, of the resource.",
```

**Review by mtdowling — CHANGES_REQUESTED** [review]

**Review by srchase — APPROVED** [review]

**Pull request merged**